addons/binding/org.openhab.binding.solaredge/src/main/java/org/openhab/binding/solaredge/internal/connector/WebInterface.java

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.solaredge.internal.connector;
+
+import java.io.UnsupportedEncodingException;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.http.HttpStatus.Code;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.openhab.binding.solaredge.config.SolarEdgeConfiguration;
+import org.openhab.binding.solaredge.handler.SolarEdgeHandler;
+import org.openhab.binding.solaredge.internal.command.PrivateApiTokenCheck;
+import org.openhab.binding.solaredge.internal.command.PublicApiKeyCheck;
+import org.openhab.binding.solaredge.internal.command.SolarEdgeCommand;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The connector is responsible for communication with the solaredge webportal
+ *
+ * @author Alexander Friese - initial contribution
+ */
+public class WebInterface {
+
+    private static final long PUBLIC_API_DAY_LIMIT = 300;
+    private static final long MINUTES_PER_DAY = 1440;
+    private static final long LOGIN_WAIT_TIME = 1000;
+
+    private final Logger logger = LoggerFactory.getLogger(WebInterface.class);
+
+    /**
+     * Configuration of the bridge from
+     * {@link org.openhab.BoxHandler.fritzaha.handler.FritzAhaBridgeHandler}
+     */
+    private final SolarEdgeConfiguration config;
+
+    /**
+     * Bridge thing handler for updating thing status
+     */
+    private final SolarEdgeHandler handler;
+
+    /**
+     * holds authentication status
+     */
+    private boolean authenticated = false;
+
+    /**
+     * HTTP client for asynchronous calls
+     */
+    private HttpClient asyncclient;
+    /**
+     * Maximum number of simultaneous asynchronous connections
+     */
+    private int asyncmaxconns = 20;
+
+    /**
+     * Constructor to set up interface
+     *
+     * @param config Bridge configuration
+     */
+    public WebInterface(SolarEdgeConfiguration config, SolarEdgeHandler handler) {
+        this.config = config;
+        this.handler = handler;
+        asyncclient = new HttpClient(new SslContextFactory(true));
```

Comment:
Jetty HttpClients create pretty huge thread pools - for this reason, we have a shared client instance in the framework, see [here](https://www.eclipse.org/smarthome/documentation/development/bindings/dependencies.html#for-http-operations). This instance can be obtained by having HttpClientFactory injected in the HandlerFactory and then pass it through the constructors to where it is used. Would it be possible to use this here or is there any strong reason, why a separate client is required?

Reply:
I did not know that there is a shared instance. Of course I could use it. In the referred page there is only the hint that the client should be obtained via "HttpClientFactory" but that class is not known in my project. What am I doing wrong?

Reply:
You are right, the documentation is hardly existing for that :-(
You need [this package import](https://github.com/eclipse/smarthome/pull/5126/files#diff-aa43a42f64f37cd5f6589f30e2bea252R23), then you can have it [injected in your handler factory](https://github.com/eclipse/smarthome/pull/5126/files#diff-2e37bb117262a383ce29b6928135e3d3R61) and then you can [pass the httpClient to the handler in the constructor](https://github.com/eclipse/smarthome/pull/5126/files#diff-2e37bb117262a383ce29b6928135e3d3R54). I hope this helps to get it sorted out!

Reply:
My Eclipse cannot resolve the class "org.eclipse.smarthome.io.net.http.HttpClientFactory". Is it necessary to setup the IDE from scatch to get these new??? packages available?

The same applies to the "org.eclipse.smarthome.core.library.types.QuantityType". The yahoowheather project does not compile within my Eclipse because the class/package is unknown.

Reply:
Sounds as if you have to refresh the target platform.
Open the file "openhab.target" with the Target platform editor, select the line `http://download.eclipse.org/smarthome/updates-stable/` and click "Reload". Once it is done, select "Reload target platform" in the upper right corner of the editor.

Reply:
Well it ended in setting up a new IDE because I wrecked it...
However now the classes are found. But I am still confused how this "injection" works.

I think this code is responsible for doing it?

    @Reference
    protected void setHttpClientFactory(HttpClientFactory httpClientFactory) {
        logger.debug("setHttpClientFactory");
        // this.httpClient = httpClientFactory.getCommonHttpClient();
    }

    protected void unsetHttpClientFactory(HttpClientFactory httpClientFactory) {
        logger.debug("unsetHttpClientFactory");
        // this.httpClient = null;
    }

But it is not called upon binding start so I guess there is some configuration missing that tells the runtime to look within my class for dependencies to inject. At least I know that Spring needs a hint where to resolve dependencies.

Reply:
Correct, because only the factory is an OSGi DS component (marked by the `@Component` annotation on the class).

Reply:
unfortunately it seems that the shared http client does not allow this:

        if (config.isUsePrivateApi()) {
            // token cookie is only used by private API therefore this can be skipped when using public API
            CookieStore cookieStore = asyncclient.getCookieStore();
            HttpCookie c = new HttpCookie(PRIVATE_API_TOKEN_COOKIE_NAME, config.getTokenOrApiKey());
            c.setDomain(PRIVATE_API_TOKEN_COOKIE_DOMAIN);
            c.setPath(PRIVATE_API_TOKEN_COOKIE_PATH);
            cookieStore.add(URI.create(getURL()), c);
        } else {